src/java/org/apache/cassandra/index/sai/utils/PrimaryKey.java

```diff
@@ -0,0 +1,184 @@
+                                                               : clusteringComparator.asByteComparable(clustering)
+                                                                                     .asComparableBytes(version);
+        return ByteSource.withTerminator(version == ByteComparable.Version.LEGACY
+                                         ? ByteSource.END_OF_STREAM
+                                         : ByteSource.TERMINATOR,
+                                         tokenComparable,
+                                         keyComparable,
+                                         clusteringComparable);
+    }
+
+    @Override
+    public int compareTo(PrimaryKey o)
+    {
+        int cmp = token().compareTo(o.token());
+
+        // If the tokens don't match then we don't need to compare any more of the key.
+        // Otherwise if it's partition key is null or the other partition key is null
+        // then one or both of the keys are token only so we can only compare tokens
+        if ((cmp != 0) || (partitionKey == null) || o.partitionKey() == null)
+            return cmp;
+
+        // Next compare the partition keys. If they are not equal or
+        // this is a single row partition key or there are no
+        // clusterings then we can return the result of this without
+        // needing to compare the clusterings
+        cmp = partitionKey().compareTo(o.partitionKey());
+        if (cmp != 0 || hasEmptyClustering() || o.hasEmptyClustering())
+            return cmp;
+        return clusteringComparator.compare(clustering(), o.clustering());
+    }
+
+    /**
+     * Return whether the primary key has an empty clustering or not.
+     * By default the clustering is empty if the internal clustering
+     * is null or is empty.
+     *
+     * @return {@code true} if the clustering is empty, otherwise {@code false}
+     */
+    public boolean hasEmptyClustering()
+    {
+        return clustering() == null || clustering().isEmpty();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(token, partitionKey, clustering, clusteringComparator);
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (obj instanceof PrimaryKey)
+            return compareTo((PrimaryKey)obj) == 0;
+        return false;
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("RowAwarePrimaryKey: { token: %s, partition: %s, clustering: %s:%s} ",
+                             token,
+                             partitionKey,
+                             clustering == null ? null : clustering.kind(),
+                             clustering == null ? null :String.join(",", Arrays.stream(clustering.getBufferArray())
+                                                                               .map(ByteBufferUtil::bytesToHex)
+                                                                               .collect(Collectors.toList())));
```

Comment:
Nit: we can use a `joining` collector:
```suggestion
                             clustering == null ? null : Arrays.stream(clustering.getBufferArray())
                                                               .map(ByteBufferUtil::bytesToHex)
                                                               .collect(Collectors.joining(",")));
```